### PR TITLE
Add pass that calls upstream datalayout propagation patterns

### PR DESCRIPTION
**Pull request by nirvedhmeshram**
This pass is not currently in a pipeline. But just adding it if we want it when and if we enable fusion.

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/CMakeLists.txt [hunk 1]
```diff
@@ -44,6 +44,7 @@ iree_cc_library(
     "AMDAIEPad.cpp"
     "AMDAIEPeelForLoop.cpp"
     "AMDAIETileAndFuse.cpp"
+    "AMDAIEPropagateDataLayout.cpp"
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/CMakeLists.txt:47]: Alphabetically needs to go one level up.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEPropagateDataLayout.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/Transforms/PassDetail.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-amdaie-propagate-pad"
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEPropagateDataLayout.cpp:13]: Can we have a common pass name?

Currently, I see that :-
DEBUG_TYPE = `iree-amdaie-propagate-pad`
.td = `iree-amdaie-propagate-data-layout`
Test file name = `datalayout_propagation`

It'd be nice to have a common name - helps debugging/navigating eventually.

**Review by Abhishek-Varma — CHANGES_REQUESTED** [review]
This is nice. But a few comments to address before getting this in.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEPropagateDataLayout.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  AMDAIEPropagateDataLayoutPass() = default;
+  AMDAIEPropagateDataLayoutPass(const AMDAIEPropagateDataLayoutPass &pass){};
+  void runOnOperation() override;
+};
+
+void AMDAIEPropagateDataLayoutPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  RewritePatternSet patterns(context);
+  linalg::populateDataLayoutPropagationPatterns(
+      patterns, [](Operation *op) { return true; });
```
Comment (MaheshRavishankar) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEPropagateDataLayout.cpp:35]: Do we want to just propagate everything by default.... We also need to call this from multiple places. Its ok to land and iterate, but this probably needs a bit more control.

**Review by MaheshRavishankar — APPROVED** [review]

**Review by Abhishek-Varma — APPROVED** [review]
Thanks! LGTM!

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEPropagateDataLayout.cpp (hunk 3 above)
Comment (nirvedhmeshram) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEPropagateDataLayout.cpp:35]: Ya it works for the test case I tried but may have to tweak based on what we see in the wild. Will keep this in mind.

**Pull request merged**